### PR TITLE
[Reference Only/Draft] joint_trajectory_controller convert trajectories to effort values

### PR DESCRIPTION
**Pull request by ddengster**
Used in conjunction with https://github.com/ros-simulation/gazebo_ros_pkgs/pull/1083

The code changes in this PR was intended to work with a 'coffeebot' simulation in gazebo. It has lots of hacks, and is only meant to be an incomplete guide to porting to ROS2.

Things of note will be put in reviews.

Update: Also used with moveit2 training https://github.com/ros-planning/moveit2/pull/187 with change 9a493a5

### CODE REVIEW

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp [hunk 1]
```diff
@@ -85,31 +90,154 @@ JointTrajectoryController::update()
...
+  {
+    state_desired = traj_msg_home_ptr_->points[0];
 
-  publish_state();
+    for (auto index = 0ul; index < joint_num; ++index) {
+        compute_error_for_joint(state_error, index, state_current, state_desired);
+    }
+  }
+  
+  double period = lifecycle_node_->get_parameter("period").as_double();
```
Comment (ddengster) [inline joint_trajectory_controller/src/joint_trajectory_controller.cpp:226]: This hack is supplied upstream from gazebo_ros_controls; ideally we should use something else.

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp [hunk 2]
```diff
@@ -85,31 +90,154 @@ JointTrajectoryController::update()
...
+  //convert to an effort command buffer via ClosedLoopHardwareInterfaceAdapter::updateCommand
+  for (auto index = 0ul; index < joint_num; ++index) {
```
Comment (ddengster) [inline joint_trajectory_controller/src/joint_trajectory_controller.cpp:228]: I use control_toolbox::pid to convert the errors to effort values.

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp [hunk 3]
```diff
@@ -253,6 +397,62 @@ JointTrajectoryController::on_configure(const rclcpp_lifecycle::State & previous
...
+  {
+    /*lifecycle_node_->declare_parameter("gains." + joint_names_[j] + ".p");
+    lifecycle_node_->declare_parameter("gains." + joint_names_[j] + ".d");
+    lifecycle_node_->declare_parameter("gains." + joint_names_[j] + ".i");
+    lifecycle_node_->declare_parameter("gains." + joint_names_[j] + ".i_clamp");*/
+
+    double p = lifecycle_node_->get_parameter("gains." + joint_names_[j] + ".p").get_value<double>();
+    double d = lifecycle_node_->get_parameter("gains." + joint_names_[j] + ".d").get_value<double>();
+    double i = lifecycle_node_->get_parameter("gains." + joint_names_[j] + ".i").get_value<double>();
+    double iclamp = lifecycle_node_->get_parameter("gains." + joint_names_[j] + ".i_clamp").get_value<double>();
```
Comment (ddengster) [inline joint_trajectory_controller/src/joint_trajectory_controller.cpp:447]: pid parameters are retrieved from the node.

File: joint_trajectory_controller/src/joint_trajectory_controller.cpp [hunk 4]
```diff
@@ -85,31 +90,154 @@ JointTrajectoryController::update()
...
+  //convert to an effort command buffer via ClosedLoopHardwareInterfaceAdapter::updateCommand
+  for (auto index = 0ul; index < joint_num; ++index) {
+    
+    double f = 0.0;
+    if (index < state_error.positions.size() && index < state_error.velocities.size() && index < state_desired.velocities.size())
+    {
+      f = pids_[index]->computeCommand(state_error.positions[index], state_error.velocities[index], rclcpp::Duration::from_seconds(period));
+
+      const double command = (state_desired.velocities[index] * velocity_ff_[index]) + f;
+      registered_joint_cmd_handles_[index]->set_cmd(command);
```
Comment (ddengster) [inline joint_trajectory_controller/src/joint_trajectory_controller.cpp:236]: It took a significant amount of effort tracing down effort_controllers::JointTrajectoryController. But it boiled down to this code. I use control_toolbox::pid to convert the errors to effort values.

**Comment by bmagyar** [comment]
This PR can now be rebased (I've squashed the jtc PR, here you can do an interactive rebase and drop everything up to `removed declareSegmentTolerances function `

**Comment by BasVolkers** [comment]
Hi what is the status of this pull request? 

I'm looking for a ROS2 replacement for `effort_controllers/JointTrajectoryController`, and I'm getting the feeling this pull request may containing what I'm looking for.

**Pull request closed without merging**